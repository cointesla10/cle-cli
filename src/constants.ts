import type { UserConfig } from './config'

export const DEFAULT_CONFIG: Required<UserConfig> = {
  JsonRpcProviderUrl: {
    mainnet: 'https://rpc.ankr.com/eth',
    sepolia: 'https://rpc.ankr.com/eth_sepolia',
    goerli: '',
  },

  UserPrivateKey: '',

  /**
   * @deprecated Please use ProverProviderUrl instead
   */
  ZkwasmProviderUrl: 'https://rpc.zkwasmhub.com:8090',

  ProverProviderUrl: 'https://rpc.zkwasmhub.com:8090',
  CompilerServerEndpoint: 'http://compiler.ora.io/compile',
  PinataEndpoint: 'https://api.pinata.cloud/pinning/pinFileToIPFS',
  PinataJWT: '',

  WasmBinPath: '[root]/build/cle.wasm',

  YamlPath: '[root]/src/cle.yaml',

  OutputProofFilePath: '[root]/build/proof_[taskId].txt',

  logger: {
    level: 'info',
  },
}

export const TAGS = {
  root: process.cwd(),
}

export const NETWORKS = [
  {
    name: 'Sepolia',
    label: 'Sepolia',
    value: 11155111,
    expectedEth: 0.002,
    hex: '0xaa36a7',
  },
  {
    name: 'Goerli',
    label: 'Goerli',
    value: 5,
    expectedEth: 0.5,
    hex: '0x5',
  },
  {
    name: 'Mainnet',
    label: 'Mainnet',
    value: 1,
  },
]

export const TdConfig = {
  fee: '0.005',
  contract: '0x47C7FC91bf15b2Af9075f55Ae0F57ce18b1328B3',
  queryrApi: 'https://zkwasm.hyperoracle.io/td',
  providerUrl: 'https://ethereum-sepolia.publicnode.com',
}

export const DspStaticParamsMap = {
  prove: 'proveParams',
  exec: 'execParams',
}
