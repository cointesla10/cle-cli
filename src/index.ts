export { run } from './cli'
export { defineConfig } from './config'
export type { UserConfig } from './config'
