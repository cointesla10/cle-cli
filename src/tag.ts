/**
 * e.g. [root]build/xxx.wasm
 * parse to project root path/xxx.wasm
 * @param template
 * @param tags
 * @returns
 */
export function parseTemplateTag(template: string, tags: Record<string, string>) {
  const value = template.replace(/\[(\w+)\]/g, (_, key) => {
    const isExistTag = Reflect.has(tags, key)
    const value = isExistTag ? tags[key] : `[${key}]`
    return value
  })
  return value
}
