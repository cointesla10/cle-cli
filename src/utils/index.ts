export * from './utils'
export * from './load'
export * from './log'
export * from './network'
export * from './path'
export * from './cli'
export * from './pinata'
export * from './file'
export * from './dsp'
