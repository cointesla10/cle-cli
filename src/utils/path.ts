import path from 'node:path'
import fs from 'node:fs'

export function createOnNonexist(filePath: string): void {
  const directoryPath = path.dirname(filePath)

  if (!fs.existsSync(directoryPath))
    fs.mkdirSync(directoryPath, { recursive: true })
}
