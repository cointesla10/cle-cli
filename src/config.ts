import { createConfigLoader } from 'unconfig'
import { DEFAULT_CONFIG, TAGS } from './constants'
import { parseTemplateTag } from './tag'
import type { LogLevel } from './logger'

export interface UserConfig {
  /**
   * Update your Etherum JSON RPC provider URL here.
   * Recommend using Erigon node RPC.
   */
  JsonRpcProviderUrl?: {
    mainnet?: string // Optional
    sepolia?: string // Optional
    goerli?: string // Optional
  }
  /**
   * Update your private key here to sign zkwasm messages.
   * Please note that (during testnet phrase) your address balance (in zkwasm server) should > 0;
   */
  UserPrivateKey?: string

  /**
   * @default "https://rpc.zkwasmhub.com:8090"
   */
  ZkwasmProviderUrl?: string
  /**
   * @default "http://compiler.dev.hyperoracle.io/compile"
   */
  CompilerServerEndpoint?: string
  /**
   * @default "https://api.pinata.cloud/pinning/pinFileToIPFS"
   */
  PinataEndpoint?: string
  PinataJWT?: string

  /**
   * @default "[root]/build/cle.wasm"
   */
  WasmBinPath?: string

  /**
   * @default "[root]/src/cle.yaml"
   */
  YamlPath?: string

  /**
   * @default "[root]/build/proof_[taskId].txt"
   */
  OutputProofFilePath?: string

  logger?: {
    /**
     * logger level
     * @default "info"
     */
    level?: LogLevel
  }
}

export type UserConfigFnObject = (...args: any[]) => UserConfig
export type UserConfigFnPromise = (...args: any[]) => Promise<UserConfig>
export type UserConfigFn = (...args: any[]) => UserConfig | Promise<UserConfig>
export type UserConfigExport = UserConfig | UserConfigFnObject | UserConfigFnPromise | Promise<UserConfig> | UserConfigFn

export function defineConfig(config: UserConfig): UserConfig
export function defineConfig(config: Promise<UserConfig>): Promise<UserConfig>
export function defineConfig(config: UserConfigFnObject): UserConfigFnObject
export function defineConfig(config: UserConfigFn): UserConfigFn
export function defineConfig(config: UserConfigExport): UserConfigExport {
  return config
}

export async function loadConfigFromFile(configFile?: string, configRoot: string = process.cwd()) {
  async function rewrite(config: any) {
    const resolved = await (typeof config === 'function' ? config() : config)
    return resolved
  }

  const loader = await createConfigLoader<UserConfig>({
    cwd: configRoot,
    sources: [configFile
      ? {
          files: configFile || '',
          extensions: [],
          rewrite: rewrite as any,
        }
      : {
          files: 'cle.config',
          extensions: ['ts', 'mts', 'cts', 'js', 'mjs', 'cjs', 'json'],
          rewrite: rewrite as any,
        }],
    defaults: DEFAULT_CONFIG,
  })
  const result = await loader.load()
  return result.config
}

export async function getConfig(configFile?: string, configRoot?: string) {
  const userConfig = await loadConfigFromFile(configFile, configRoot)

  userConfig.WasmBinPath = parseTemplateTag(userConfig.WasmBinPath || '', TAGS)
  // userConfig.LocalWasmBinPath = parseTemplateTag(userConfig.LocalWasmBinPath || '', TAGS)
  userConfig.YamlPath = parseTemplateTag(userConfig.YamlPath || '', TAGS)

  return userConfig as Required<UserConfig>
}
