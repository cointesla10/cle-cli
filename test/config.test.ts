import path from 'node:path'
import { describe, expect, it } from 'vitest'
import { getConfig, loadConfigFromFile } from '../src/config'

const configFixturesRoot = path.join(__dirname, 'fixtures/config')

describe('config', () => {
  it('load default config', async () => {
    const config = await loadConfigFromFile(undefined, configFixturesRoot)
    expect(config).toMatchSnapshot()
  })

  it('load function config', async () => {
    const funcConfigPath = 'cle.func.config.ts'
    const config = await loadConfigFromFile(funcConfigPath, configFixturesRoot)
    expect(config).toMatchSnapshot()
  })

  it('merge config', async () => {
    const config = await getConfig('cle.merge.config.ts', configFixturesRoot) as any
    // its value will change at runtime
    // so in order to pass the test correctly
    // delete it.
    delete config.LocalWasmBinPath
    delete config.YamlPath
    delete config.MappingPath
    expect(config).toMatchSnapshot()
  })

  it('load object config', async () => {
    const funcConfigPath = 'cle.object.config.ts'
    const config = await loadConfigFromFile(funcConfigPath, configFixturesRoot)
    expect(config).toMatchSnapshot()
  })
})
