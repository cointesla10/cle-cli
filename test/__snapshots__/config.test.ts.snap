// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`config > load default config 1`] = `
{
  "CompilerServerEndpoint": "http://compiler.ora.io/compile",
  "JsonRpcProviderUrl": {
    "goerli": "https://eth-goerli.nodereal.io/v1/5c1f0a30d6ff4093a31afce579bca8d6",
    "mainnet": "https://eth-mainnet.nodereal.io/v1/5c1f0a30d6ff4093a31afce579bca8d6",
    "sepolia": "https://rpc.ankr.com/eth_sepolia",
  },
  "OutputProofFilePath": "[root]/build/proof_[taskId].txt",
  "PinataEndpoint": "https://api.pinata.cloud/pinning/pinFileToIPFS",
  "PinataJWT": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VySW5mb3JtYXRpb24iOnsiaWQiOiJhNmQyY2VmMS03ODMzLTQ0YWEtYjdjMi1jYmJmMGUyOTA2OWYiLCJlbWFpbCI6InN5Mjk0MEBueXUuZWR1IiwiZW1haWxfdmVyaWZpZWQiOnRydWUsInBpbl9wb2xpY3kiOnsicmVnaW9ucyI6W3siaWQiOiJGUkExIiwiZGVzaXJlZFJlcGxpY2F0aW9uQ291bnQiOjF9LHsiaWQiOiJOWUMxIiwiZGVzaXJlZFJlcGxpY2F0aW9uQ291bnQiOjF9XSwidmVyc2lvbiI6MX0sIm1mYV9lbmFibGVkIjpmYWxzZSwic3RhdHVzIjoiQUNUSVZFIn0sImF1dGhlbnRpY2F0aW9uVHlwZSI6InNjb3BlZEtleSIsInNjb3BlZEtleUtleSI6ImY2NmY0ZDE2ZjhjZTQxMjEzY2YyIiwic2NvcGVkS2V5U2VjcmV0IjoiZjcyZjcwMGJjNDkzYTIzMGNmNjAxMmFkOTQzZDY2MjIzMzkyZTJiZjI5NTE5MzNjMzNhOGVhYjU0NTM1NzE3OCIsImlhdCI6MTY5MTgwOTUyNH0.OU-IY0AV1OABmtX3ySBBdzvvrIgz8IGCRbgOBPvRO0Y",
  "ProverProviderUrl": "https://rpc.zkwasmhub.com:8090",
  "UserPrivateKey": "0x0",
  "WasmBinPath": "build/cle.wasm",
  "YamlPath": "[root]/src/cle.yaml",
  "ZkwasmProviderUrl": "https://zkwasm-explorer.delphinuslab.com:8090",
  "logger": {
    "level": "info",
  },
}
`;

exports[`config > load function config 1`] = `
{
  "CompilerServerEndpoint": "http://compiler.ora.io/compile",
  "JsonRpcProviderUrl": {
    "goerli": "https://eth-goerli.nodereal.io/v1/5c1f0a30d6ff4093a31afce579bca8d6",
    "mainnet": "https://eth-mainnet.nodereal.io/v1/5c1f0a30d6ff4093a31afce579bca8d6",
    "sepolia": "https://rpc.ankr.com/eth_sepolia",
  },
  "OutputProofFilePath": "[root]/build/proof_[taskId].txt",
  "PinataEndpoint": "https://api.pinata.cloud/pinning/pinFileToIPFS",
  "PinataJWT": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VySW5mb3JtYXRpb24iOnsiaWQiOiJhNmQyY2VmMS03ODMzLTQ0YWEtYjdjMi1jYmJmMGUyOTA2OWYiLCJlbWFpbCI6InN5Mjk0MEBueXUuZWR1IiwiZW1haWxfdmVyaWZpZWQiOnRydWUsInBpbl9wb2xpY3kiOnsicmVnaW9ucyI6W3siaWQiOiJGUkExIiwiZGVzaXJlZFJlcGxpY2F0aW9uQ291bnQiOjF9LHsiaWQiOiJOWUMxIiwiZGVzaXJlZFJlcGxpY2F0aW9uQ291bnQiOjF9XSwidmVyc2lvbiI6MX0sIm1mYV9lbmFibGVkIjpmYWxzZSwic3RhdHVzIjoiQUNUSVZFIn0sImF1dGhlbnRpY2F0aW9uVHlwZSI6InNjb3BlZEtleSIsInNjb3BlZEtleUtleSI6ImY2NmY0ZDE2ZjhjZTQxMjEzY2YyIiwic2NvcGVkS2V5U2VjcmV0IjoiZjcyZjcwMGJjNDkzYTIzMGNmNjAxMmFkOTQzZDY2MjIzMzkyZTJiZjI5NTE5MzNjMzNhOGVhYjU0NTM1NzE3OCIsImlhdCI6MTY5MTgwOTUyNH0.OU-IY0AV1OABmtX3ySBBdzvvrIgz8IGCRbgOBPvRO0Y",
  "ProverProviderUrl": "https://rpc.zkwasmhub.com:8090",
  "UserPrivateKey": "0x0",
  "WasmBinPath": "build/cle.wasm",
  "YamlPath": "[root]/src/cle.yaml",
  "ZkwasmProviderUrl": "https://zkwasm-explorer.delphinuslab.com:8090",
  "logger": {
    "level": "info",
  },
}
`;

exports[`config > load object config 1`] = `
{
  "CompilerServerEndpoint": "http://compiler.ora.io/compile",
  "JsonRpcProviderUrl": {
    "goerli": "https://eth-goerli.nodereal.io/v1/5c1f0a30d6ff4093a31afce579bca8d6",
    "mainnet": "https://eth-mainnet.nodereal.io/v1/5c1f0a30d6ff4093a31afce579bca8d6",
    "sepolia": "https://rpc.ankr.com/eth_sepolia",
  },
  "OutputProofFilePath": "[root]/build/proof_[taskId].txt",
  "PinataEndpoint": "https://api.pinata.cloud/pinning/pinFileToIPFS",
  "PinataJWT": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VySW5mb3JtYXRpb24iOnsiaWQiOiJhNmQyY2VmMS03ODMzLTQ0YWEtYjdjMi1jYmJmMGUyOTA2OWYiLCJlbWFpbCI6InN5Mjk0MEBueXUuZWR1IiwiZW1haWxfdmVyaWZpZWQiOnRydWUsInBpbl9wb2xpY3kiOnsicmVnaW9ucyI6W3siaWQiOiJGUkExIiwiZGVzaXJlZFJlcGxpY2F0aW9uQ291bnQiOjF9LHsiaWQiOiJOWUMxIiwiZGVzaXJlZFJlcGxpY2F0aW9uQ291bnQiOjF9XSwidmVyc2lvbiI6MX0sIm1mYV9lbmFibGVkIjpmYWxzZSwic3RhdHVzIjoiQUNUSVZFIn0sImF1dGhlbnRpY2F0aW9uVHlwZSI6InNjb3BlZEtleSIsInNjb3BlZEtleUtleSI6ImY2NmY0ZDE2ZjhjZTQxMjEzY2YyIiwic2NvcGVkS2V5U2VjcmV0IjoiZjcyZjcwMGJjNDkzYTIzMGNmNjAxMmFkOTQzZDY2MjIzMzkyZTJiZjI5NTE5MzNjMzNhOGVhYjU0NTM1NzE3OCIsImlhdCI6MTY5MTgwOTUyNH0.OU-IY0AV1OABmtX3ySBBdzvvrIgz8IGCRbgOBPvRO0Y",
  "ProverProviderUrl": "https://rpc.zkwasmhub.com:8090",
  "UserPrivateKey": "0x0",
  "WasmBinPath": "build/cle.wasm",
  "YamlPath": "[root]/src/cle.yaml",
  "ZkwasmProviderUrl": "https://zkwasm-explorer.delphinuslab.com:8090",
  "logger": {
    "level": "info",
  },
}
`;

exports[`config > merge config 1`] = `
{
  "CompilerServerEndpoint": "https://compiler.ora.io/compile/",
  "JsonRpcProviderUrl": {
    "goerli": "https://eth-goerli.nodereal.io/v1/5c1f0a30d6ff4093a31afce579bca8d6",
    "mainnet": "https://eth-mainnet.nodereal.io/v1/5c1f0a30d6ff4093a31afce579bca8d6",
    "sepolia": "https://rpc.ankr.com/eth_sepolia",
  },
  "OutputProofFilePath": "[root]/build/proof_[taskId].txt",
  "PinataEndpoint": "https://api.pinata.cloud/pinning/pinFileToIPFS",
  "PinataJWT": "PinataJWT",
  "ProverProviderUrl": "https://rpc.zkwasmhub.com:8090",
  "UserPrivateKey": "UserPrivateKey",
  "WasmBinPath": "WasmBinPath",
  "ZkwasmProviderUrl": "https://rpc.zkwasmhub.com:8090",
  "logger": {
    "level": "info",
  },
}
`;
